hack/deployer/runner/k3d.go

```diff
@@ -0,0 +1,204 @@
+}
+
+func (k *K3dDriver) delete() error {
+	// The cluster delete operation will not properly remove the kubeconfig entry
+	// as the /home directory isn't mounted in the k3d flow.
+	cmd := k.cmd("cluster", "delete")
+	if cmd == nil {
+		return fmt.Errorf("failed to create k3d cluster")
+	}
+	if err := cmd.Run(); err != nil {
+		return err
+	}
+	// Manually remove the kubeconfig entry.
+	clusterName := fmt.Sprintf("%s-%s", "k3d", k.plan.ClusterName)
+	userName := fmt.Sprintf("admin@%s", clusterName)
+	// ignore errors when removing clusters/users/contexts using k3d
+	_ = removeKubeconfig(clusterName, clusterName, userName)
+	return nil
+}
+
+func (k *K3dDriver) cmd(args ...string) *exec.Command {
+	params := map[string]interface{}{
+		"ClusterName":    k.plan.ClusterName,
+		"SharedVolume":   env.SharedVolumeName(),
+		"K3dClientImage": k.clientImage,
+		"K3dNodeImage":   k.nodeImage,
+		"Args":           args,
+	}
+
+	dockerSocket := "/var/run/docker.sock"
+
+	var socketExists bool
+	_, err := os.Stat(dockerSocket)
+	socketExists = !os.IsNotExist(err)
+
+	// If we are on macOS and the docker socket does not exist, we need to
+	// fall back to using the docker socket in the user's home directory
+	// as with recent changes for docker desktop /var/run/docker.sock
+	// can be created/used, and the docker socket in $HOME errors.
+	if runtime.GOOS == "darwin" && !socketExists {
+		dockerSocket = "$HOME/.docker/run/docker.sock"
+	}
+	// We need the docker socket so that k3d can bootstrap
+	// --userns=host to support Docker daemon host configured to run containers only in user namespaces
+	command := `docker run --rm \
+		--userns=host \
+		-v /var/run/docker.sock:` + dockerSocket + ` \
+		-e HOME=/home \
+		-e PATH=/ \
+		{{.K3dClientImage}} \
+		{{Join .Args " "}} {{.ClusterName}}`
+	return exec.NewCommand(command).AsTemplate(params)
+}
+
+func (k *K3dDriver) getKubeConfig() (*os.File, error) {
+	// Get kubeconfig from k3d binary
+	output, err := k.cmd("kubeconfig", "get", k.plan.ClusterName).WithoutStreaming().Output()
+	if err != nil {
+		return nil, err
+	}
+
+	// Replace host.docker.internal with 127.0.0.1 only if not running inside the CI container
+	// and on macOS.
+	if os.Getenv("CI") != "true" && runtime.GOOS == "darwin" {
+		output = strings.ReplaceAll(output, "host.docker.internal", "127.0.0.1")
+	}
+
+	// Persist kubeconfig for reliability in following kubectl commands
+	kubeCfg, err := os.CreateTemp("", "kubeconfig")
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = kubeCfg.Write([]byte(output))
+	if err != nil {
+		return nil, err
+	}
+	return kubeCfg, nil
+}
+
+func (k *K3dDriver) GetCredentials() error {
+	config, err := k.getKubeConfig()
+	if err != nil {
+		return err
+	}
+	defer os.Remove(config.Name())
+	return mergeKubeconfig(config.Name())
+}
+
+func (k *K3dDriver) createTmpStorageClass() (string, error) {
+	tmpFile := filepath.Join(os.Getenv("HOME"), storageClassFileName)
```

Comment:
Let's maybe use a temporary directory instead of `$HOME` and ensure that this file is eventually deleted?